src/java/org/apache/cassandra/io/util/FileUtils.java

```diff
@@ -67,23 +71,84 @@
     public static final boolean isCleanerAvailable;
     private static final AtomicReference<Optional<FSErrorHandler>> fsErrorHandler = new AtomicReference<>(Optional.empty());
 
+    private static Class clsDirectBuffer;
+    private static MethodHandle mhDirectBufferCleaner;
+    private static MethodHandle mhCleanerClean;
+    private static MethodHandle mhDirectBufferAddress;
+
     static
     {
         boolean canClean = false;
         try
         {
+            clsDirectBuffer = Class.forName("sun.nio.ch.DirectBuffer");
+            Method mDirectBufferCleaner = clsDirectBuffer.getMethod("cleaner");
+            mhDirectBufferCleaner = MethodHandles.lookup().unreflect(mDirectBufferCleaner);
+            Method mCleanerClean = mDirectBufferCleaner.getReturnType().getMethod("clean");
+            mhCleanerClean = MethodHandles.lookup().unreflect(mCleanerClean);
+            Method mDirectBufferAddress = clsDirectBuffer.getMethod("address");
+            mhDirectBufferAddress = MethodHandles.lookup().unreflect(mDirectBufferAddress);
+
             ByteBuffer buf = ByteBuffer.allocateDirect(1);
-            ((DirectBuffer) buf).cleaner().clean();
+            cleanerClean(buf);
             canClean = true;
         }
         catch (Throwable t)
         {
+            logger.info("Cannot initialize un-mmaper. Compacted data files will not be removed promptly.", t);
             JVMStabilityInspector.inspectThrowable(t);
-            logger.info("Cannot initialize un-mmaper.  (Are you using a non-Oracle JVM?)  Compacted data files will not be removed promptly.  Consider using an Oracle JVM or using standard disk access mode");
         }
         isCleanerAvailable = canClean;
     }
 
+    public static boolean isDirectBuffer(ByteBuffer buf)
+    {
+        return clsDirectBuffer.isInstance(buf);
+    }
+
+    public static void cleanerClean(ByteBuffer buf)
+    {
+        cleanerClean(buf, false);
+    }
+
+    public static void cleanerClean(ByteBuffer buf, boolean cleanAttachment)
```

Comment:
I don't see any code path where `cleanAttachment` can be true. Am I missing something?

Reply:
yikes - that was wrong. good catch!